common/djangoapps/dark_lang/api.py

```diff
@@ -0,0 +1,199 @@
+CHINESE_LANGUAGE_CODE_MAP = {
+    # The following are the new-style language codes for chinese language
+    'zh-hans': 'zh-CN',  # Chinese (Simplified),
+    'zh-hans-cn': 'zh-CN',  # Chinese (Simplified, China)
+    'zh-hans-sg': 'zh-CN',  # Chinese (Simplified, Singapore)
+    'zh-hant': 'zh-TW',  # Chinese (Traditional)
+    'zh-hant-hk': 'zh-HK',  # Chinese (Traditional, Hongkong)
+    'zh-hant-mo': 'zh-TW',  # Chinese (Traditional, Macau)
+    'zh-hant-tw': 'zh-TW',  # Chinese (Traditional, Taiwan)
+    # The following are the old-style language codes that django does not recognize
+    'zh-mo': 'zh-TW',  # Chinese (Traditional, Macau)
+    'zh-sg': 'zh-CN',  # Chinese (Simplified, Singapore)
+}
+LANGUAGE_INPUT_FIELD = 'preview_lang'
+RESET_SUBMIT = 'reset'
+SET_LANGUAGE_SUBMIT = 'set_language'
+
+
+class Darklang(object):
+    """
+    Support class and code for dark-launching languages.
+
+    This is configured by creating ``DarkLangConfig`` rows in the database,
+    using the django admin site.
+    """
+
+    @property
+    def released_langs(self):
+        """
+        Current list of released languages
+        """
+        language_options = DarkLangConfig.current().released_languages_list
+        if settings.LANGUAGE_CODE not in language_options:
+            language_options.append(settings.LANGUAGE_CODE)
+        return language_options
+
+    def process_darklang_request(self, request):
+        """
+        Prevent user from requesting un-released languages
+        """
+        if not DarkLangConfig.current().enabled:
+            return None
+
+        if RESET_SUBMIT in request.POST:
+            # Reset and clear the language preference
+            return 'Language reset to %s' % self.clear_preview_language(request)
+        if SET_LANGUAGE_SUBMIT in request.POST:
+            # Set the Preview Language
+            return self.set_preview_language(request)
+
+    def set_preview_language(self, request):
+        """
+        Attempt to set the Preview language
+        :param request: The Django Request Object
+        :return dark language setting code
+        """
+        if not DarkLangConfig.current().enabled:
+            return None
+
+        if LANGUAGE_INPUT_FIELD not in request.POST:
+            return None
+
+        preview_lang = request.POST[LANGUAGE_INPUT_FIELD]
+
+        if preview_lang == '':
+            return None
+
+        auth_user = request.user.is_authenticated()
+
+        # Set the session key to the requested preview lang
+        request.session[LANGUAGE_SESSION_KEY] = preview_lang
+
+        # Make sure that we set the requested preview lang as the dark lang preference for the
+        # user, so that the lang_pref middleware doesn't clobber away the dark lang preview.
+        if auth_user:
+            set_user_preference(request.user, DARK_LANGUAGE_KEY, preview_lang)
+        return 'Language set to %s' % preview_lang
```

Comment:
For consistency with clear_preview_lang, how about this method just returns preview_lang?